server/src/main/java/org/elasticsearch/common/util/Maps.java

```diff
@@ -279,6 +280,25 @@ public static <K, V> LinkedHashMap<K, V> newLinkedHashMapWithExpectedSize(int ex
         return new LinkedHashMap<>(capacity(expectedSize));
     }
 
+    /**
+     * Returns the sub-map of the given {@code map} that contains all the entries under keys with the given {@code prefix}.
+     *
+     * @param map sorted map
+     * @param prefix prefix to filter by
+     * @return map
+     */
+    public static <V> SortedMap<String, V> subMap(SortedMap<String, V> map, String prefix) {
+        // create the the next prefix right after the given prefix, and use it as exclusive upper bound for the sub-map to filter by prefix
+        // below
+        if (prefix.isEmpty()) {
+            return map;
+        }
+        char[] toPrefixCharArr = prefix.toCharArray();
+        toPrefixCharArr[toPrefixCharArr.length - 1]++;
```

Comment:
I know this code is as before, but there's a bug here (although very unlikely). If for some wild reason `toPrefixCharArr[toPrefixCharArr.length - 1]` was `\uFFFF`, the + 1 would wrap around and the subMap will return an incorrect result. An if/else with this condition and a prefix with an extra character of `\u0000` should handle this edge case.

I think we should fix this since the method is now part of the generic Maps class.